pkg/provider/provider.go

```diff
@@ -359,14 +361,127 @@ func (k *kubeProvider) Invoke(ctx context.Context,
 
 	// Unmarshal arguments.
 	tok := req.GetTok()
+	label := fmt.Sprintf("%s.Invoke(%s)", k.label(), tok)
+	args, err := plugin.UnmarshalProperties(
+		req.GetArgs(), plugin.MarshalOptions{Label: label, KeepUnknowns: true})
+	if err != nil {
+		return nil, pkgerrors.Wrapf(err, "failed to unmarshal %v args during an Invoke call", tok)
+	}
 
 	// Process Invoke call.
 	switch tok {
+	case invokeList:
+		namespace := ""
+		if args["namespace"].HasValue() {
+			namespace = args["namespace"].StringValue()
+		}
+		if !args["group"].HasValue() || !args["version"].HasValue() || !args["kind"].HasValue() {
+			return nil, fmt.Errorf(
+				"watch requires a group, version, and kind that uniquely specify the resource type")
+		}
+		cl, err := k.clientSet.ResourceClient(schema.GroupVersionKind{
+			Group:   args["group"].StringValue(),
+			Version: args["version"].StringValue(),
+			Kind:    args["kind"].StringValue(),
+		}, namespace)
+		if err != nil {
+			return nil, err
+		}
+
+		list, err := cl.List(metav1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		objects := []map[string]interface{}{}
+		for _, o := range list.Items {
+			objects = append(objects, o.Object)
+		}
+		resp, err := plugin.MarshalProperties(
+			resource.NewPropertyMapFromMap(map[string]interface{}{"items": objects}),
+			plugin.MarshalOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		return &pulumirpc.InvokeResponse{Return: resp}, nil
 	default:
 		return nil, fmt.Errorf("Unknown Invoke type '%s'", tok)
 	}
 }
 
+// Invoke dynamically executes a built-in function in the provider.
+func (k *kubeProvider) StreamInvoke(req *pulumirpc.InvokeRequest, server pulumirpc.ResourceProvider_StreamInvokeServer) error {
+	// Unmarshal arguments.
+	tok := req.GetTok()
+	label := fmt.Sprintf("%s.StreamInvoke(%s)", k.label(), tok)
+	args, err := plugin.UnmarshalProperties(
+		req.GetArgs(), plugin.MarshalOptions{Label: label, KeepUnknowns: true})
+	if err != nil {
+		return pkgerrors.Wrapf(err, "failed to unmarshal %v args during an Invoke call", tok)
+	}
+
+	// Process StreamInvoke call.
+	switch tok {
+	case streamInvokeWatch:
+		//
+		// Set up resource watcher.
+		//
+
+		namespace := ""
+		if args["namespace"].HasValue() {
+			namespace = args["namespace"].StringValue()
+		}
+		if !args["group"].HasValue() || !args["version"].HasValue() || !args["kind"].HasValue() {
+			return fmt.Errorf(
+				"watch requires a group, version, and kind that uniquely specify the resource type")
+		}
+		cl, err := k.clientSet.ResourceClient(schema.GroupVersionKind{
+			Group:   args["group"].StringValue(),
+			Version: args["version"].StringValue(),
+			Kind:    args["kind"].StringValue(),
+		}, namespace)
+		if err != nil {
+			return err
+		}
+
+		watch, err := cl.Watch(metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+
+		//
+		// Channel to signals `StreamInvoke` is complete. Usually ctrl-c.
```

Comment:
Does this comment only apply to `k.canceler.context.Done()`?

Reply:
Oops. When I removed the `watch` implementation I accidentally blew away the accompanying changes I made to the provider. There's actually an entire clause missing here, which I'll need to rewrite.